### ISSUE TITLE
YOLOv8 TFLite INT8 Export Issue: Full Integer Quantized Model Fails to Detect Objects

### ISSUE THREAD

**Issue by hzidou** [ultralytics/ultralytics#21628]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I export yolov8n.pt to TFLite INT8 with Ultralytics:
```
from ultralytics import YOLO

# Load a model
model = YOLO(runs/detect/train/weights/best.pt")  # load a custom trained model

# Export the model
model.export(format="TFLite",int8=True,data="dataset/data.yaml",imgsz=640)
```
generates followings:
├── assets
├── variables
├── best_float16.tflite
├── best_float32.tflite
├── best_full_integer_quant.tflite
├── best_integer_quant.tflite
├── fingerprint.pb
├── metadata.yaml
├── saved_model.pb

I use examples/YOLOv8-TFLite-Python/main.py
best_float16.tflite best_float32.tflite is worked but best_full_integer_quant.tflite failed

Then，I try to use the following code:
```
from ultralytics import YOLO

model_path="best_saved_model/best_full_integer_quant.tflite"
file="hd12.jpg"
conf=0.25

model = YOLO(model_path)
results = model.predict(source = file, save=True, conf=conf, show_labels=True, show_conf=True, save_txt=True, device=0)
```

Output:
```
image 1/1 F:\ultralytics\ultralytics-main\hd12.jpg: 640x640 (no detections), 51.7ms
Speed: 27.8ms preprocess, 51.7ms inference, 25.2ms postprocess per image at shape (1, 3, 640, 640)
Results saved to runs\detect\predict3
0 label saved to runs\detect\predict3\labels
```
As the output shown, the full integer quantized model produces incorrect results.

<img width="632" height="674" alt="Image" src="https://github.com/user-attachments/assets/53333d9d-08fb-4a35-bef2-0c63e49c6a35" />

I suspect the quantization of output is wrong .
I would really appreciate if anyone can help me find the issue.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @hzidou, thank you for opening this issue and for using Ultralytics 🚀 This is an automated response to help you get support faster. An Ultralytics engineer will follow up with you shortly.

We recommend a visit to the Docs for helpful guides and examples:
- Python usage: https://docs.ultralytics.com/usage/python/
- CLI usage: https://docs.ultralytics.com/usage/cli/
- General Docs: https://docs.ultralytics.com/

If this is a 🐛 Bug Report, please provide a minimum reproducible example (MRE) to help us debug it:
- https://docs.ultralytics.com/help/minimum-reproducible-example/

If this is a custom training ❓ Question, please include:
- Dataset details and a few sample images
- Training/export logs and exact commands
- Confirmation you’ve reviewed Tips for Best Training Results: https://docs.ultralytics.com/guides/model-training-tips/

Join the Ultralytics community where it suits you best:
- Discord for real-time chat 🎧 https://discord.com/invite/ultralytics
- Discourse for in-depth discussions https://community.ultralytics.com/
- Subreddit to share knowledge https://www.reddit.com/r/Ultralytics/

## Upgrade

Please upgrade to the latest ultralytics package with all requirements in a Python>=3.8 environment with PyTorch>=1.8 to verify the issue is not already resolved in the latest version:
```bash
pip install -U ultralytics
```
Requirements: https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml

## Environments

YOLO can be run in the following up-to-date verified environments (all dependencies including CUDA/CUDNN, Python and PyTorch preinstalled):

- Notebooks with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- Google Cloud Deep Learning VM. See GCP Quickstart Guide: https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/
- Amazon Deep Learning AMI. See AWS Quickstart Guide: https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/
- Docker Image. See Docker Quickstart Guide: https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/ <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all Ultralytics CI tests are passing. CI tests verify correct operation of all YOLO Modes and Tasks on macOS, Windows, and Ubuntu every 24 hours and on every commit.

To help us assist you faster with the TFLite INT8 export behavior you described, please:
1.Share a minimal script showing your exact export command and calibration data setup, plus the smallest image(s) that reproduce the issue
2.Include the exported model list you shared and confirm ultralytics and dependencies versions
3.Provide the prediction command output for float32/float16 vs full-integer models and the sample image you used

Thanks again for the detailed report and logs — we’ll take a closer look soon 🔍

**Comment by Y-T-G** [comment]
Can you provide the output after running this command in terminal: `yolo checks`?

**Comment by hzidou** [comment]
> Can you provide the output after running this command in terminal: `yolo checks`?

I am exporting on a Linux system
```
Ultralytics 8.3.173 🚀 Python-3.10.12 torch-2.5.0+cu118 CUDA:0 (NVIDIA A10, 24030MiB)
Setup complete ✅ (8 CPUs, 28.9 GB RAM, 37.8/97.9 GB disk)

OS                  Linux-5.10.134-18.0.2.lifsea8.x86_64-x86_64-with-glibc2.35
Environment         Linux
Python              3.10.12
Install             pip
Path                /nasmnt/envs/cosyvoice_vllm_train/lib/python3.10/site-packages/ultralytics
RAM                 28.91 GB
Disk                37.8/97.9 GB
CPU                 Intel Xeon Platinum 8369B 2.90GHz
CPU count           8
GPU                 NVIDIA A10, 24030MiB
GPU count           1
CUDA                11.8

numpy               ✅ 1.26.4>=1.23.0
matplotlib          ✅ 3.7.5>=3.3.0
opencv-python       ✅ 4.11.0.86>=4.6.0
pillow              ✅ 10.4.0>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.14.1>=1.4.1
torch               ✅ 2.5.0+cu118>=1.8.0
torch               ✅ 2.5.0+cu118!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.20.0+cu118>=0.9.0
tqdm                ✅ 4.67.1>=4.64.0
psutil              ✅ 6.1.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.3>=1.1.4
ultralytics-thop    ✅ 2.0.14>=2.0.0
```

Predicting on Windows system
```
Ultralytics 8.3.176  Python-3.10.18 torch-2.5.0+cu118 CUDA:0 (NVIDIA GeForce RTX 3060 Laptop GPU, 6144MiB)
Setup complete  (16 CPUs, 15.7 GB RAM, 727.5/931.5 GB disk)

OS                  Windows-10-10.0.19045-SP0
Environment         Windows
Python              3.10.18
Install             pip
Path                F:\anaconda\envs\ultralytics\Lib\site-packages\ultralytics
RAM                 15.74 GB
Disk                727.5/931.5 GB
CPU                 11th Gen Intel Core(TM) i7-11800H 2.30GHz
CPU count           16
GPU                 NVIDIA GeForce RTX 3060 Laptop GPU, 6144MiB
GPU count           1
CUDA                11.8

numpy                2.1.3>=1.23.0
matplotlib           3.10.3>=3.3.0
opencv-python        4.12.0.88>=4.6.0
pillow               11.3.0>=7.1.2
pyyaml               6.0.2>=5.3.1
requests             2.32.4>=2.23.0
scipy                1.15.3>=1.4.1
torch                2.5.0+cu118>=1.8.0
torch                2.5.0+cu118!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision          0.20.0+cu118>=0.9.0
tqdm                 4.67.1>=4.64.0
psutil               7.0.0
py-cpuinfo           9.0.0
pandas               2.3.1>=1.1.4
ultralytics-thop     2.0.14>=2.0.0
```

**Comment by Y-T-G** [comment]
I can't reproduce the issue. The prediction works fine with full integer quant model. Did you test the prediction on the linux system?

**Comment by hzidou** [comment]
> I can't reproduce the issue. The prediction works fine with full integer quant model. Did you test the prediction on the linux system?

The prediction of Linux system is also not feasible.

But I found that the model structure of yolov8n_full_integer_quant.tflite is different from my best_full_integer_quant.tflite model structure.
And the scale of yolov8d_full_integer_quant.tflite's output is 0.003906319383531809, while the scale of my output is 2.6477792623031006

**yolov8n_full_integer_quant.tflite:**
<img width="412" height="803" alt="Image" src="https://github.com/user-attachments/assets/f1eb5cd8-1496-40d4-a38d-ff0f68c8a270" />

**my tflite:**
<img width="387" height="763" alt="Image" src="https://github.com/user-attachments/assets/8147aa7e-e17f-4a25-bee9-604507421b8e" />

I am very confused about the different model structures because I have not made any modifications to the yolov8n model.

**Comment by Y-T-G** [comment]
Is the prediction of YOLOv8n full integer quant model working?

**Comment by hzidou** [comment]
> Is the prediction of YOLOv8n full integer quant model working?

I'm sorry for the late reply. Yes, the YOLOv8n full integer quant model works well.

**Comment by glenn-jocher** [comment]
Thanks, that helps narrow it down to your custom weights/export. Please re-export with the latest Ultralytics and the pinned converter stack (onnx2tf ≥1.26.3, ai-edge-litert 1.2–<1.4, onnx <1.18) used by our exporter, ensure imgsz matches training, and calibrate with a representative set via data=... (consider fraction=1.0); then sanity-check inference with a low conf (e.g., 0.001). If it still fails, attach the full export log from the “TensorFlow SavedModel” step and the failing .tflite’s metadata.yaml so we can reproduce. For reference, see the TFLite export guide with INT8 args and calibration details and the exporter SavedModel step that lists required package versions.

**Comment by hzidou** [comment]
Thank you so much for your help! I tried re-exporting the model using Google Colab, and even though I’m not exactly sure what the original issue was, the exported TFLite model is now successfully making predictions.
Thank you again for your assistance—your guidance made all the difference!

**Comment by glenn-jocher** [comment]
Great to hear it’s working now! These INT8 TFLite mismatches are often caused by converter stack drift, so exporting in a clean, pinned environment (as you did on Colab) is the right approach; if all is resolved, please close the issue, and if it recurs, attach the export log and the failing .tflite’s metadata.yaml so we can reproduce.
